dataset_loaders/images/gta5.py

```diff
@@ -0,0 +1,290 @@
+            21: (107, 142, 35),     # vegetation
+            22: (152, 251, 152),    # terrain
+            23: (0, 130, 180),      # sky
+            24: (220, 20, 60),      # person
+            25: (255, 0, 0),        # rider
+            26: (0, 0, 142),        # car
+            27: (0, 0, 70),         # truck
+            28: (0, 60, 100),       # bus
+            29: (0,  0, 0),         # caravan
+            30: (0,  0, 0),         # trailer
+            31: (0, 80, 100),       # train
+            32: (0, 0, 230),        # motorcycle
+            33: (119, 11, 32),      # bicycle
+            -1: (0, 0, 0)         # license plate
+            # 5: (111, 74, 0),        # dynamic
+            # 6: (81,  0, 81),        # ground
+            # 9: (250, 170, 160),     # parking
+            # 10: (230, 150, 140),    # rail track
+            # 14: (180, 165, 180),    # guard rail
+            # 15: (150, 100, 100),    # bridge
+            # 16: (150, 120, 90),     # tunnel
+            # 18: (153, 153, 153),    # polegroup
+            # 29: (0,  0, 90),        # caravan
+            # 30: (0,  0, 110),       # trailer
+            }
+
+    _filenames = None
+    _prefix_list = None
+
+    @property
+    def prefix_list(self):
+        if self._prefix_list is None:
+            # Create a list of prefix out of the number of requested videos
+            self._prefix_list = np.unique(np.array([el[:6]
+                                                    for el in self.filenames]))
+
+        return self._prefix_list
+
+    @property
+    def filenames(self):
+        if self._filenames is None:
+            # Get file names for this set
+            filenames = []
+            import scipy.io
+            split = scipy.io.loadmat(os.path.join(self.path, 'split.mat'))
+            split = split[self.which_set + "Ids"]
+            for id in split:
+                filenames.append(str(id[0]).zfill(5)+'.png')
+            self._filenames = filenames
+            # print(filenames)
+        return self._filenames
+
+    def __init__(self, which_set='train', *args, **kwargs):
+
+        self.which_set = "val" if which_set == "valid" else which_set
+        self.image_path = os.path.join(self.path, "images")
+        self.mask_path = os.path.join(self.path, "labels")
+
+        # constructing the ThreadedDataset
+        # it also creates/copies the dataset in self.path if not already there
+        super(GTA5Dataset, self).__init__(*args, **kwargs)
+
+    def get_names(self):
+        """Return a dict of names, per prefix/subset."""
+        per_subset_names = {}
+        # Populate self.filenames and self.prefix_list
+        filenames = self.filenames
+        prefix_list = self.prefix_list
+
+        # cycle through the different videos
+        for prefix in prefix_list:
+            per_subset_names[prefix] = [el for el in filenames if
+                                        el.startswith(prefix)]
```

Comment:
Filtering the filenames by the prefix every time `get_names` is called seems wasteful: why don't you make `self.filenames` a dictionary and modify the filenames function to store a list of filenames for each prefix?

This should work but please check it:
```python
def filenames(self):
    if self._filenames is None:
        filenames = {}
        split = scipy.io.loadmat(os.path.join(self.path, 'split.mat'))
        split = split[self.which_set + "Ids"]
        for name in split:
            prefix = name[:6]
            filenames.setdefault(prefix, []).append(name[0].zfill(5) + '.png')
        self._filenames = filenames
    return self._filenames
```

You can then remove `self.prefix_list` and the related function, and return self.filenames directly in `get_names()`.